Sources/General/Commands/Setup.swift

```diff
@@ -0,0 +1,337 @@
+    }
+
+    private func downloadArchive(at path: String) throws -> URL {
+        guard let url = URL(string: path) else {
+            throw Error.url(path)
+        }
+        let data = try self.data(at: url)
+        let tmpFolderURL = try tempFolderWithRandomName()
+        let fileURL = tmpFolderURL + url.lastPathComponent
+        do {
+            try data.write(to: fileURL)
+        }
+        catch {
+            throw Error.write(tmpFolderURL)
+        }
+        return fileURL
+    }
+
+    private func data(at url: URL) throws -> Data {
+        do {
+            return try Data(contentsOf: url)
+        }
+        catch {
+            throw Error.download(url)
+        }
+    }
+
+    private func tempFolderWithRandomName() throws -> URL {
+        let root = fileManager.homeDirectoryForCurrentUser
+        let url = root + Constants.appFolderName + "tmp" + UUID().uuidString
+        do {
+            try fileManager.createDirectory(at: url, withIntermediateDirectories: true, attributes: nil)
+        }
+        catch {
+            throw Error.write(url)
+        }
+        return url
+    }
+
+    private func unzipArchive(at url: URL) throws -> URL {
+        let folderURL = url.deletingLastPathComponent()
+        do {
+            try fileManager.unzipItem(at: url, to: folderURL)
+        }
+        catch {
+            throw Error.write(folderURL)
+        }
+        return folderURL
+    }
+
+    private func loadSteupFiles(in url: URL) -> (templates: [URL], spec: URL?) {
+        var templates = [URL]()
+        var spec: URL?
+        do {
+            let contents = try fileManager.contentsOfDirectory(at: url,
+                                                               includingPropertiesForKeys: nil,
+                                                               options: [])
+            for url in contents {
+                var isDirectory: ObjCBool = false
+                _ = fileManager.fileExists(atPath: url.path, isDirectory: &isDirectory)
+                if isDirectory.boolValue {
+                    if fileManager.fileExists(atPath: (url + Constants.specFilename).path) {
+                        templates.append(url)
+                    }
+                    else {
+                        let setupFiles = loadSteupFiles(in: url)
+                        templates.append(contentsOf: setupFiles.templates)
+                        if let specURL = setupFiles.spec {
+                            spec = specURL
+                        }
+                    }
+                }
+                else if url.lastPathComponent == Constants.generalSpecName {
+                    spec = url
+                }
+                else if url.pathExtension == "stencil" {
+                    templates.append(url.deletingLastPathComponent())
+                }
+            }
+        }
+        catch {
+            return (Array(Set(templates)), spec)
+        }
+        return (Array(Set(templates)), spec)
```

Comment:
Since the catch body is the same as normal return, just doing a guard on getting contents of the directory should make everything a bit cleaner.

Reply:
is not actual now